src/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Keychain.cs

```diff
@@ -0,0 +1,327 @@
+        }
+
+        internal static SafeTemporaryKeychainHandle CreateTemporaryKeychain()
+        {
+            string tmpKeychainPath = Path.Combine(
+                Path.GetTempPath(),
+                Guid.NewGuid().ToString("N") + ".keychain");
+
+            // Use a distinct GUID so that if a keychain is abandoned it isn't recoverable.
+            string tmpKeychainPassphrase = Guid.NewGuid().ToString("N");
+
+            byte[] utf8Passphrase = System.Text.Encoding.UTF8.GetBytes(tmpKeychainPassphrase);
+
+            SafeTemporaryKeychainHandle keychain;
+
+            int osStatus = AppleCryptoNative_SecKeychainCreateTemporary(
+                tmpKeychainPath,
+                utf8Passphrase.Length,
+                utf8Passphrase,
+                out keychain);
+
+            SafeTemporaryKeychainHandle.TrackKeychain(keychain);
+
+            if (osStatus != 0)
+            {
+                keychain.Dispose();
+                throw CreateExceptionForOSStatus(osStatus);
+            }
+
+            return keychain;
+        }
+
+        internal static void SecKeychainDelete(IntPtr handle, bool throwOnError=true)
+        {
+            int osStatus = AppleCryptoNative_SecKeychainDelete(handle);
+
+            if (throwOnError && osStatus != 0)
+            {
+                throw CreateExceptionForOSStatus(osStatus);
+            }
+        }
+    }
+}
+
+namespace System.Security.Cryptography.Apple
+{
+    internal class SafeKeychainItemHandle : SafeHandle
+    {
+        internal SafeKeychainItemHandle()
+            : base(IntPtr.Zero, ownsHandle: true)
+        {
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            SafeTemporaryKeychainHandle.ReleaseItem(handle);
+            Interop.CoreFoundation.CFRelease(handle);
+            SetHandle(IntPtr.Zero);
+            return true;
+        }
+
+        public override bool IsInvalid => handle == IntPtr.Zero;
+    }
+
+    internal class SafeKeychainHandle : SafeHandle
+    {
+        internal SafeKeychainHandle()
+            : base(IntPtr.Zero, ownsHandle: true)
+        {
+        }
+
+        internal SafeKeychainHandle(IntPtr handle)
+            : base(handle, ownsHandle: true)
+        {
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Interop.CoreFoundation.CFRelease(handle);
+            SetHandle(IntPtr.Zero);
+            return true;
+        }
+
+        public override bool IsInvalid => handle == IntPtr.Zero;
+    }
+
+    internal sealed class SafeTemporaryKeychainHandle : SafeKeychainHandle
+    {
+        private static readonly Dictionary<IntPtr, SafeTemporaryKeychainHandle> s_lookup =
+            new Dictionary<IntPtr, SafeTemporaryKeychainHandle>();
+
+        internal SafeTemporaryKeychainHandle()
+        {
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            lock (s_lookup)
+            {
+                s_lookup.Remove(handle);
```

Comment:
If you mean managed (DangerousAddRef()), the count is 0 when ReleaseHandle is called.

If you mean native, that's tracked by the CFRelease/CFRetain (and I've seen that a keychain being deleted while someone has a handle to it results in OSStatus errors, not crashes)